src/ng/parse.js

```diff
@@ -622,15 +622,44 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isPure(node, parentIsPure) {
+  switch (node.type) {
+    // Computed members might invoke a stateful toString()
+    case AST.MemberExpression:
+      if (node.computed) {
+        return false;
+      }
+      break;
+
+    // Unary always convert to primative
+    case AST.UnaryExpression:
+      return true;
```

Comment:
First of all, I don't know what `&&` and `||` have to do with this (since they are not unary operators) :confused:

Regarding `getValueOf()`, I don't think it is enough to "save" us (in cases where `valueOf()` returns non-primitive values). I admit it is very unlikely that this will affect any real app, but from a theoretical point of view, I think expressions such as `'+obj'`, where `obj` has a `valueOf()` method that returns an object will not be handled correctly (e.g. if the returned object stays the same by reference, but one of its properties changes).